org.eclipse.jdt.core/model/org/eclipse/jdt/core/compiler/CompilationParticipant.java

```diff
@@ -152,6 +152,31 @@ public void processAnnotations(BuildContext[] files) {
 	// do nothing by default
 }
 
+/**
+ * Returns where to find annotation processors. If this is not specified,
+ * then the class path is searched for processors.
+ *
+ * @param project the project to participate in
+ * @param isTest whether the annotation processor path is for test code
+ * @return the annotation processor paths
+ * @since 3.38
```

Comment:
Would be better if such API changes could be avoided, given that is more like an experiment

Reply:
Annotation processing configurations, such as the processor path and output directory, are part of the standard compiler arguments. It is better to explicitly provide these configurations within the host builder (for instance, AbstractImageBuilder) when invoking the custom compiler.

Currently, jdt.core utilizes the CompilationParticipant mechanism to interact with the APT plugin for annotation processing tasks. Extending the CompilationParticipant class appears to be a natural choice for obtaining APT configurations. Any other ideas we could consider for accessing APT information?

Reply:
> Any other ideas we could consider for accessing APT information?

Does your client code need to call or override that API or can jdt just typecast to its internal implementation?

Reply:
> Does your client code need to call or override that API

My client code does not use this API. It's AbstractImageBuilder in jdt.core that calls the API to get the related information and pass the result to the client code.

> can jdt just typecast to its internal implementation?

jdt.core doesn't depend on the apt plugin directly. It uses extension point to discover AptCompilationParticipant instance. Not sure how to use typecast.

Reply:
Let a new intermediate jdt internal class 
AbstractAptCompilationParticipant extends CompilationParticipant, let 
AptCompilationParticipant extends AbstractAptCompilationParticipant (jdt apt should be a friend of jdt), 
downcast CompilationParticipant to AbstractAptCompilationParticipant if the participiant is an instance of that.  
That's a little detour but avoids otherwise unneeded public API.
 -- Or mark the new API as Provisional so that it could be removed again without further warning.

Reply:
Good point. thanks for suggestion.